src/uri/serializer/microuriserializer.rs

```diff
@@ -84,8 +84,10 @@ impl UriSerializer<Vec<u8>> for MicroUriSerializer {
     /// A `Vec<u8>` representing the serialized `UUri`.
     #[allow(clippy::cast_possible_truncation)]
     fn serialize(uri: &UUri) -> Result<Vec<u8>, SerializationError> {
-        if UriValidator::is_empty(uri) || !UriValidator::is_micro_form(uri) {
-            return Err(SerializationError::new("URI is empty or not in micro form"));
+        if let Err(validation_error) = UriValidator::validate_micro_form(uri) {
```

Comment:
now that we return an Error we might also want to add some doc comment and maybe some examples (which can also serve as test cases)?

Reply:
How might that look?

Currently have several test cases that I added to check the new functionality [down here](https://github.com/eclipse-uprotocol/up-rust/pull/18/files/daba0f5647c3923eb1385069d3b78daefeefda70#diff-200aba31079074af7b1567bf5eee9492883a3da6087d7a31d97bacf95a932209R580-R650).

Reply:
See https://doc.rust-lang.org/rustdoc/write-documentation/documentation-tests.html

Reply:
So you wanna see doc tests where, over on `validate_micro_form()` inside of `uurivalidator.rs`?

Or here on `serialize()`?

I think I need you to explain a bit more what you're looking for.

Reply:
In the doc comment on the `serialize` function you could provide example code that creates a UUri, then serializes it and asserts the byte array content is as expected. That would basically be sufficient to replace the `test_serialize_uri` unit test ...
Maybe you can find some other unit tests that could serve as usage examples for the other functions as well.

Reply:
Ahhh, I think I get it. You'd like to improve the docs a bit on `serialize` with some doc test examples, right?

Perhaps show 1 x success and 1 x failure?

Wanna confirm before I run off and do it :slightly_smiling_face:

Reply:
Exactly, and while you're at it, maybe also improve the docs of some other functions as well :-)

Reply:
And ideally, we can then remove some of the unit tests because they are now implemented as doc tests ...

Reply:
All right -- gave this a go [here](https://github.com/eclipse-uprotocol/up-rust/pull/18/files#diff-50d5281e2de4d6b2de3ffb2294e4a0bb1fe84a474bd1595b080e740cba9e2aacR202-R334) and [here](https://github.com/eclipse-uprotocol/up-rust/pull/18/files#diff-200aba31079074af7b1567bf5eee9492883a3da6087d7a31d97bacf95a932209R91-R143).

Looking for any feedback you might have :slightly_smiling_face: 

BTW -- I discussed with @stevenhartley that there's a bit of ambiguity in how the id or ip or both or neither are valid states for UAuthority at the moment, so we should wait on merging this till he has it resolved in the .proto. I'll have to revise a bit based on his change to the .proto.